### PR TITLE
[MultiModal] Fusion model inference acceleration with TensorRT

### PR DESCRIPTION
**Pull request by liangfu**
*Issue #, if available:*

*Description of changes:*

This PR brings support to accelerate fusion models with tensorrt.

For a fusion model trained with petfinder dataset, TensorrtExecutionProvider can boost inference speed up to 2.7x faster, comparing to GPU-based realtime prediction with PyTorch.

```
"petfinder",
["numerical_mlp", "categorical_mlp", "timm_image", "hf_text", "fusion_mlp"],
"google/electra-small-discriminator",
"mobilenetv3_small_100"
```

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by tonyhoo** [comment]
Thanks for the change @liangfu. Do you know why preprocessing logics are generally slower in the TRT execution env?

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 1]
```diff
@@ -73,82 +76,71 @@ def dump_model(self, save_path: Optional[str] = None):
...
+        path
+            The export path of onnx model.
         batch_size
             The batch_size of export model's input.
             Normally the batch_size is a dynamic axis, so we could use a small value for faster export.
         verbose
             verbose flag in torch.onnx.export.
         opset_version
             opset_version flag in torch.onnx.export.
+        truncate_long_and_double: bool, default False
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/utils/export.py:103]: Do we need to do special handing as well for `bf16`?

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 2]
```diff
@@ -73,82 +76,71 @@ def dump_model(self, save_path: Optional[str] = None):
...
-                data=data,
-                valid_input=valid_input,
-                onnx_tracing=True,
-                batch_size=batch_size,
-            )
-
-        if not onnx_path:
-            onnx_path = os.path.join(self.path, default_onnx_path)
-
-        device_type = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (tonyhoo) [inline multimodal/src/autogluon/multimodal/utils/export.py:129]: Wondering why previously we use GPU for `onnx_export`

**Review by tonyhoo — COMMENTED** [review]
Can you add some accuracy metrics as well just to make sure accuracy doesn't compromise post TRT transformation

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 1 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:103]: It's handled in the TensorrtExecutionProvider configuration.
```python
"TensorrtExecutionProvider",
{
    "device_id": 0,
    "trt_max_workspace_size": 2147483648,
    "trt_fp16_enable": True,
},
```

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:103]: truncate_long_and_double here just to convert double -> float, int64 -> int32, since TRT wouldn't support int64 and double data types at all.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 2 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:129]: We actually used cpu instead, see deleted comment below.

To quote
```
# Perform tracing on cpu, since we're facing an error when tracing with cuda device:
#     ERROR: Tensor-valued Constant nodes differed in value across invocations.
#     This often indicates that the tracer has encountered untraceable code.
#     Comparison exception:   The values for attribute 'shape' do not match: torch.Size([]) != torch.Size([384]).
#     from https://github.com/rwightman/pytorch-image-models/blob/3aa31f537d5fbf6be8f1aaf5a36f6bbb4a55a726/timm/models/swin_transformer.py#L112
device = "cpu"
num_gpus = 0
```

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    test_df = dataset.test_df.head(batch_size)
+    tail_df = dataset.test_df.tail(batch_size)
+
+    # Prediction with default predictor
+    y_pred = predictor.predict(test_df)
+
+    trt_module = predictor.export_tensorrt(path=model_path, data=tail_df, batch_size=batch_size)
+
+    # To use the TensorRT module for prediction, simply replace the _model in the predictor
+    predictor._model = trt_module
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_deployment_tensorrt.py:84]: This design seems different from onnx, which uses `MultiModalOnnxPredictor`. Do we need to unify them?

File: multimodal/src/autogluon/multimodal/models/clip.py [hunk 4]
```diff
@@ -169,6 +169,7 @@ def forward(
                 return_dict=True,
             )
             text_features = self.model.text_projection(text_outputs.pooler_output)  # (b, num_features)
+            last_hidden_state = text_outputs[0]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/clip.py:172]: Does `text_outputs.last_hidden_state` not work for TensorRT?

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/clip.py:172]: This has to be changed, since we are adding the `torchscript=True` argument in 
```
AutoConfig.from_pretrained(checkpoint_name, torchscript=True)
```

Basically, `text_outputs[0]` (with `torchscript=True`) equals to `text_ouputs.last_hidden_state` (with `torchscript=False`).

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py (hunk 3 above)
Comment (liangfu) [reply multimodal/tests/unittests/others/test_deployment_tensorrt.py:84]: Yes, I totally agree. I think `simply replace the _model in the predictor` would be better for integration, comparing to the way we were integrating with MultiModalPredictor.

But shall we put the unifying effort in a follow-up PR? It might be too overwhelm to deprecate too many old implementation in a single PR.

**Comment by liangfu** [comment]
> Thanks for the change @liangfu. Do you know why preprocessing logics are generally slower in the TRT execution env?

Good question, I'm also interested to see why this is happening. Therefore, I did a breakdown of the call stack

image

Seems like data_processor is a bit slower when using TRT.

File: multimodal/src/autogluon/multimodal/models/fusion/fusion_mlp.py [hunk 5]
```diff
@@ -119,20 +119,29 @@ def __init__(
         self.name_to_id = self.get_layer_ids()
         self.head_layer_names = [n for n, layer_id in self.name_to_id.items() if layer_id == 0]
 
+    @property
+    def input_keys(self):
+        module_list = self.model
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/fusion/fusion_mlp.py:124]: This line seems unused?

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py [hunk 6]
```diff
@@ -219,7 +219,8 @@ def forward(
         }
         if text_column_names:
             assert len(text_column_names) == len(text_column_indices), "invalid text column inputs"
-            batch.update(**dict(zip(text_column_names, text_column_indices)))
+            for idx, name in enumerate(text_column_names):
+                batch[name] = text_column_indices[idx]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/huggingface_text.py:223]: Is this change necessary?

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:223]: Yes, tracing can't unpack dict.

File: multimodal/src/autogluon/multimodal/models/ner_text.py [hunk 7]
```diff
@@ -164,11 +164,11 @@ def forward(
                 attention_mask=text_masks,
             )
 
-        sequence_output = outputs.last_hidden_state
+        sequence_output = outputs[0]  # last_hidden_state
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/ner_text.py:167]: How about using `last_hidden_state = outputs[0]` to be consistent with other models?

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/ner_text.py]: Agree. changed sequence_output -> last_hidden_state.

File: multimodal/src/autogluon/multimodal/models/fusion/fusion_mlp.py (hunk 5 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/fusion/fusion_mlp.py]: Nice catch. This line has been removed.

File: multimodal/src/autogluon/multimodal/models/huggingface_text.py (hunk 6 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/huggingface_text.py:223]: Yes, they have the same effect.

File: multimodal/src/autogluon/multimodal/models/t_few.py [hunk 8]
```diff
@@ -234,25 +245,51 @@ def forward(
...
+            return features, logits, target_template_logits, lm_target, column_features, column_feature_masks
+
+    def get_output_dict(
+        self,
+        features: torch.Tensor,
+        logits: torch.Tensor,
+        target_template_logits: torch.Tensor,
+        lm_target: torch.Tensor,
+        column_features: Optional[List[torch.Tensor]] = None,
+        column_feature_masks: Optional[List[torch.Tensor]] = None,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/t_few.py:280]: These types should be `Optional[Dict[str, torch.Tensor]]`. Also need to correct the related arguments in other models' `get_output_dict `.

File: multimodal/src/autogluon/multimodal/models/utils.py [hunk 9]
```diff
@@ -558,10 +558,10 @@ def get_hf_config_and_model(
     -------
     A Huggingface config and model.
     """
-    config = AutoConfig.from_pretrained(checkpoint_name)
+    config = AutoConfig.from_pretrained(checkpoint_name, torchscript=True)
 
     if pretrained:
-        model = AutoModel.from_pretrained(checkpoint_name, low_cpu_mem_usage=low_cpu_mem_usage)
+        model = AutoModel.from_pretrained(checkpoint_name, low_cpu_mem_usage=low_cpu_mem_usage, torchscript=True)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/utils.py:564]: Setting `torchscript=True` make all the tied weights in model decoupled, which can cause problems if finetuning the model. See https://huggingface.co/transformers/v3.0.2/torchscript.html#torchscript-flag-and-tied-weights.

File: multimodal/src/autogluon/multimodal/models/utils.py [hunk 10]
```diff
@@ -686,28 +686,45 @@ def update_mmdet_config(key, value, config):
...
+    supported_models = (
+        TimmAutoModelForImagePrediction,
+        HFAutoModelForTextPrediction,
+        MultimodalFusionMLP,
+        TFewModel,
+        OnnxModule,
+    )
     pure_model = model.module if isinstance(model, nn.DataParallel) else model
+    if isinstance(pure_model, OnnxModule):
+        for k in batch:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/utils.py:706]: What possible types can `batch[k].dtype` be when code runs here? Can we say it is always in `[torch.float32, torch.int32, torch.int64]`?

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/tests/unittests/others/test_deployment_tensorrt.py:84]: Follow-up PRs sound good.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 10 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py:706]: This is a special handling for converting the data type in token_ids. I'm not sure why the token_ids are in int32, but the inputs are required to be int64.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 9 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py:564]: Here is a list of things to assess the potential impact of this change.
1. see how many models are impacted by tied weights, especially hf_text and derived models.
2. evaulate the potential accuracy loss when decouple the layers that have tied weights
3. try to work around the torchscript argument, and try to avoid setting it to True, in order to avoid the issue
4. (less likely but possible) if the impact of tied weight is high, and impact of TRT is relatively slow, while we don't have workaround to avoid this issue, we could potentially abandon the idea of introducing the onnx/trt exporting feature.

File: multimodal/src/autogluon/multimodal/models/t_few.py (hunk 8 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/t_few.py:280]: Nice catch. BTW, i'm just curious why we didn't catch such mistake in unittests.

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py (hunk 3 above)
Comment (liangfu) [reply multimodal/tests/unittests/others/test_deployment_tensorrt.py:84]: Will be updated in PR https://github.com/autogluon/autogluon/pull/2983

File: multimodal/src/autogluon/multimodal/utils/onnx.py [hunk 11]
```diff
@@ -7,99 +7,70 @@
...
-        The predictor's pipeline.
-    config
-        The predictor's config.
+def onnx_get_dynamic_axes(input_keys: List[str]):
+    dynamic_axes = {}
+    for k in input_keys:
+        if "token_ids" in k or "segment_ids" in k:
+            dynamic_axes[k] = {0: "batch_size", 1: "seq_length"}
+        elif "valid_length" in k or k.startswith("numerical") or k.startswith("timm_image"):
+            dynamic_axes[k] = {0: "batch_size"}
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:25]: The shape of `images` in `timm_image`'s input should be `(b, n, c, h, w)`, where both b and n may be dynamic.

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 12]
```diff
@@ -160,23 +152,85 @@ def export_onnx(
...
+            onnx_path = os.path.join(path, "model.onnx")
+        dirname = os.path.dirname(os.path.abspath(onnx_path))
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
 
-        with precision_context, torch.no_grad():
-            traced_model = torch.jit.trace(self._model, input_vec)
-        torch.jit.save(traced_model, "traced_model.pt")
-        traced_model = torch.jit.load("traced_model.pt")
+        input_keys = self._model.input_keys
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:175]: Consider putting this line at the beginning of this function. I see some places above uses `self._model.input_keys`. Using one variable can make code more organized.

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 13]
```diff
@@ -160,23 +152,85 @@ def export_onnx(
...
-        input_vec = InputBatch(**batch)
+        keys_to_drop = list(set(batch.keys()).difference(set(self._model.input_keys)))
+        for k in keys_to_drop:
+            batch.pop(k)
+        # Truncate input data types for TensorRT (only support: bool, int32, half, float)
+        if truncate_long_and_double:
+            for k in batch:
+                if batch[k].dtype == torch.int64:
+                    batch[k] = batch[k].to(torch.int32)
+        input_vec = [batch[k] for k in self._model.input_keys]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:163]: Could we move all the logic to process/convert `batch` data, including the code to support mixed precision on floating point inputs, to util function `get_processed_batch_for_deployment `?  I think these logic better fits that context.

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 14]
```diff
@@ -73,82 +76,71 @@ def dump_model(self, save_path: Optional[str] = None):
 
     def export_onnx(
         self,
-        data: Optional[pd.DataFrame] = None,
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:76]: We may need to improve this example and add more examples about `timm_image`, `hf_text`, and `fusion_mlp`. Since `data` is required, I guess this example no longer works. You can mark this add a TODO item. I think tutorials may also be necessary to teach users.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 9 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/utils.py:564]: Since onnx/trt is mainly for inference, how about still using the previous logic that scripts the model immediately before exporting onnx? Setting `torchscript=True` here can impact training, which may require tied weights, especially for the generative models to add.

File: multimodal/src/autogluon/multimodal/utils/onnx.py [hunk 15]
```diff
@@ -7,99 +7,70 @@
 
 logger = logging.getLogger(__name__)
 
+try:
+    import tensorrt  # Unused but required by TensorrtExecutionProvider
+except:
+    logger.warning(
+        "Failed to import tensorrt package. "
+        "onnxruntime would fallback to CUDAExecutionProvider instead of using TensorrtExecutionProvider."
+    )
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:16]: Is it easy to install `tensorrt` for users? If not, consider making this a lazy import to avoid unnecessary warnings for users who don't need it.

File: multimodal/src/autogluon/multimodal/models/utils.py [hunk 16]
```diff
@@ -686,28 +686,45 @@ def update_mmdet_config(key, value, config):
...
-    supported_models = (TimmAutoModelForImagePrediction, HFAutoModelForTextPrediction)
+    supported_models = (
+        TimmAutoModelForImagePrediction,
+        HFAutoModelForTextPrediction,
+        MultimodalFusionMLP,
+        TFewModel,
+        OnnxModule,
+    )
     pure_model = model.module if isinstance(model, nn.DataParallel) else model
+    if isinstance(pure_model, OnnxModule):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/models/utils.py:705]: In which case `pure_model ` can be a `OnnxModule ` object? I don't see where `run_model` is called with a `OnnxModule ` object as input.

File: multimodal/src/autogluon/multimodal/utils/onnx.py [hunk 17]
```diff
@@ -7,99 +7,70 @@
...
-            },
-            "timm_image_image_valid_num": {},
-        }
-        default_onnx_path = "./model.onnx"
-        default_batch = None
-    else:
-        raise ValueError(f"ONNX export is not supported in current pipeline {pipeline}")
 
-    return valid_input, dynamic_axes, default_onnx_path, default_batch
+class OnnxModule(object):
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:30]: Consider adding some docstrings to explain this class. Tensorrt seems only one provider in onnx runtime. But it's unclear how the three providers work. Does onnxruntime check `TensorrtExecutionProvider `, `CUDAExecutionProvider `, and `CPUExecutionProvider ` sequentially to see their availability? Better to use some docstrings to describe the mechanism.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 16 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py:705]: For predicting with the TRT for a fusion model,
```python
onnx_module = predictor.export_tensorrt(data=X)
predictor._model = onnx_module
predictor.predict(X)  # pure_model would be OnnxModule here
```

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 17 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:30]: > Does onnxruntime check TensorrtExecutionProvider , CUDAExecutionProvider , and CPUExecutionProvider  sequentially to see their availability?

Yes, the execution providers are sequentially checked.

> Better to use some docstrings to describe the mechanism.

Sure.

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 15 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:16]: > Is it easy to install tensorrt for users?

Yes, it's just `pip install tensorrt`.

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:16]: tried lazy import, but didn't work.

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 18]
```diff
@@ -160,23 +152,85 @@ def export_onnx(
...
+
+        truncate_long_and_double = False
+        onnx_path = self.export_onnx(
+            data=data, path=path, batch_size=batch_size, truncate_long_and_double=truncate_long_and_double
+        )
+
+        logger.info("Loading ONNX file from path {}...".format(onnx_path))
+        onnx_model = onnx.load(onnx_path)
+
+        trt_module = OnnxModule(onnx_model)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:228]: The returned module may not use tensorrt. Maybe moving the warning of importing tensorrt inside `OnnxModule ` is better?

Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:228]: I tried to move importing tensorrt inside OnnxModule, but that didn't work. (onnxruntime.InferenceSession doesn't compile to use TensorrtExecutionProvider)

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 17 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:30]: Added docstring for describing the mechanism.

File: multimodal/src/autogluon/multimodal/models/t_few.py (hunk 8 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/t_few.py]: Corrected related arguments in T-Few model as well as other models' get_output_dict function.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 9 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py:564]: > how about still using the previous logic that scripts the model immediately before exporting onnx?

Yes, this is a better solution to satisfy both worlds. I will definitely give it a try.

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 11 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:25]: I would suggest avoid adding image size into dynamic dimensions, because
1. adding an extra dynamic dimension would increase the complexity of model compilation
2. the compiled model with too many dynamic dimensions could be suboptimal in terms of performance

The question becomes do we really need to support dynamic shape in image data? If yes, what is the lower bound and upper bound of the image size.

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 15 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:16]: It's kind of weird why lazy import doesn't work. It's fine to keep it here for now. Maybe later dive into the reason.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 14 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:76]: We might need to consider clean up these examples, since they are not tested / verified in contineous-integration. See https://github.com/autogluon/autogluon/issues/2848 

> I think tutorials may also be necessary to teach users.

I'm planning to add a tutorial for using TRT, but the tutorial would come after this PR.

File: multimodal/src/autogluon/multimodal/models/utils.py (hunk 9 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/utils.py]: torchscript=True argument has been removed, and seems everything is still working without the argument.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 13 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py]: moved to get_processed_batch_for_deployment function as suggested.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 12 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py]: placed to the top location, and the local input_keys variable is reused.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 14 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py:76]: A tutorial will be added in https://github.com/autogluon/autogluon/pull/2987

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 19]
```diff
@@ -73,118 +76,136 @@ def dump_model(self, save_path: Optional[str] = None):
...
-                data=data,
-                valid_input=valid_input,
-                onnx_tracing=True,
-                batch_size=batch_size,
-            )
+        # Infer default onnx path, and create parent directory if needed
+        if not path:
+            onnx_path = os.path.join(self.path, "model.onnx")
+        else:
+            onnx_path = os.path.join(path, "model.onnx")
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:137]: A better way may be

```
if not path:
    path = self.path
onnx_path = os.path.join(path, "model.onnx")
```

File: multimodal/src/autogluon/multimodal/models/clip.py (hunk 4 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/models/clip.py:172]: Since we don't use `torchscript=True`, do we still need `last_hidden_state = text_outputs[0]`?

File: multimodal/src/autogluon/multimodal/utils/onnx.py [hunk 20]
```diff
@@ -7,99 +7,100 @@
...
+        Returns
+        -------
+        onnx_outputs : list of torch.Tensor
+            A list of torch.Tensor that are outputs of the model.
+        """
+        import torch
+
+        input_dict = {k: args[i].cpu().numpy() for i, k in enumerate(self.input_names)}
+        onnx_outputs = self.sess.run(self.output_names, input_dict)
+        onnx_outputs = onnx_outputs[:3]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/onnx.py:95]: Why do we need only the first 3 model outputs? Don't we need them all?

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    batch_size = 2
+
+    # Subsample test data for efficient testing, and use a different subset of the dataset for compilation.
+    test_df = dataset.test_df.head(batch_size)
+    tail_df = dataset.test_df.tail(batch_size)
+
+    # Prediction with default predictor
+    y_pred = predictor.predict(test_df)
+
+    trt_module = predictor.export_tensorrt(path=model_path, data=tail_df, batch_size=batch_size)
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_deployment_tensorrt.py:81]: Currently, `export_onnx` and `export_tensorrt` either returns a path or a module. When a user calls `export_something()`, do users expect return something? Or users just expect the model is saved to disk? If users want to use the saved model, they probably want to load it?

File: multimodal/src/autogluon/multimodal/utils/export.py [hunk 22]
```diff
@@ -219,9 +240,37 @@ def get_processed_batch_for_deployment(
...
+
+        # Move model data to the specified device
+        for key in input_keys:
+            inp = batch[key]
+            # support mixed precision on floating point inputs, and leave integer inputs (for language models) untouched.
+            if inp.dtype.is_floating_point:
+                batch[key] = inp.to(device, dtype=dtype)
+            else:
+                batch[key] = inp.to(device)
+        self._model.to(device)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/export.py:262]: Do we need `self._model.to(device)` here?

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 11 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:25]: In a pretrained setting, n should be a fixed number, since input DataFrame should have same number of image columns, isn't it?

File: multimodal/src/autogluon/multimodal/utils/onnx.py (hunk 20 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/onnx.py:95]: This comes with the undetermined size of output dict for a fusion_mlp model, which contains a tuple of `(features, logits, multimodal_logits)`. The variable `multimodal_logits` contains the logit output from all modalities.

> Don't we need them all?

Good question. In short, `multimodal_logits` are only used for computing loss, not used for inference.

Specifically, `get_output_dict()` would take outputs from `forward()`, but in onnxruntime outputs are flatten, which means the list of tensors in `multimodal_logits` would be merged into other outputs (e.g. features, logits). We kind of lost the information about where did the extra tensors come from.

File: multimodal/tests/unittests/others/test_deployment_tensorrt.py (hunk 21 above)
Comment (liangfu) [reply multimodal/tests/unittests/others/test_deployment_tensorrt.py:81]: These are excellent questions.

There are several different usecases:
U1: Some users might expect an ONNX file to be exported, so that they can use the ONNX file where ever they want.
U2: Some users might expect faster inference time with ONNX. They actually don't care much about the details on where is the ONNX file.

In terms of U1, I think the expected output could be either the location of the ONNX file, or the ONNX model itself. This is the way how `export_onnx` is defined.

In terms of U2, we should be able to generate a drop-in replacement of torch.nn.Module, so that integration with existing inference flow would be easy. This is the way how `export_tensorrt` is defined.

But I would agree, we should have a better name for these public APIs.

File: multimodal/src/autogluon/multimodal/models/clip.py (hunk 4 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/models/clip.py]: undid this changed.

File: multimodal/src/autogluon/multimodal/utils/export.py (hunk 19 above)
Comment (liangfu) [reply multimodal/src/autogluon/multimodal/utils/export.py]: Updated as suggested.

**Review by zhiqiangdon — APPROVED** [review]
LGTM! Thanks for supporting Tensorrt!
